package.json

```diff
@@ -27,6 +27,7 @@
   },
   "dependencies": {
     "abstract-leveldown": "~6.2.1",
+    "buffer": "^5.5.0",
```

Comment:
I don't like this, relying on a specific version that can't be deduped between modules.

Reply:
how would you like to handle that?

Reply:
actually we can remove buffer.
whats the minimum node version level-js wants to support ?

Reply:
The dependency isn't required for Electron or browserify.

Webpack 5 no longer detects use of globals, but can it still inject dependencies?

Reply:
> whats the minimum node version level-js wants to support ?

none 😉 `level-js` is browser-only.

Reply:
electron renderer thread ran with node_integration off doesn't have buffer.
webpack 5 will not inject anything it will just mock `global`, `__dirname` and `__filename`. https://webpack.js.org/configuration/node/#nodeglobal

Reply:
lol ofc, so the question is which  browser do you want  to support? level-js only needs buffer if TextDecoder is not available.

Reply:
> level-js only needs buffer if TextDecoder is not available

It always needs buffer, because it's an `abstract-leveldown` store and the whole level ecosystem is geared towards buffers (for the time being).

Reply:
humm abstract-leveldown added this in https://github.com/Level/abstract-leveldown/pull/355 but didn't added the dependency. This seems wrong at least as a peer dependency

Reply:
I need to read up on the webpack story.

Reply:
ah abstract-leveldown isn't release yet with that change. 
IMO they are doing the right thing ppl should be able to own their dependency tree, and they had completely outdated versions of the node polyfills so basically they were doing more bad than good.

Reply:
The main argument for webpack's change doesn't apply here:

> While this \[polyfilling\] makes using modules written for node.js easy, it adds these huge polyfills to the bundle. In many cases these polyfills are unnecessary.

Because the polyfill is necessary in this case. At least until we find a way to support other binary types ecosystem-wide - but that requires much thought and likely won't happen any time soon.

Webpack also says:

> It's possible to manually add a polyfill for a node.js core module.

I think it's fair to expect that from `level-js` users, who are not just buying into `level-js` alone, but the ecosystem too (and most commonly, compatibility with `leveldown`).

Reply:
> completely outdated versions of the node polyfills

To be fair, that's because those polyfills (intended for browserify and still used by browserify enthousiasts) are maintained by a handful of people with little help.

Reply:
> I think it's fair to expect that from level-js users, who are not just buying into level-js alone, but the ecosystem too (and most commonly, compatibility with leveldown).

Exactly why i made this PR and many more in others modules we use in our code base.

> To be fair, that's because those polyfills (intended for browserify and still used by browserify enthousiasts) are maintained by a handful of people with little help.

Yeah i know, but the webpack node libs repo is completely abandoned i have PRs there opened for ages.

Reply:
> and many more in others modules we use in our code base

How do you and other module authors feel about (potentially) having different versions of core modules in the tree? Maybe I'm mistaken about how much of a problem that is.

Reply:
i have been dealing with this forever, at one point i had 4 versions of streams in my bundle (like the really old one from 0.8 through streams3) it was a nightmare. 
theres no simple solution to handle these things, we slowly fixed a million modules by helping each other, making PRs to update deps and upgrade code to newer streams behaviour.
semver if applied correctly helps but besides that its just a community effort to move things forward.

Reply:
Oh yeah, I hear you about streams.

So there's no webpack option to inject the buffer polyfill if a module imports it? Because that's how I read "it's possible to manually add a polyfill".

If not, I guess we don't have a choice 😃. And a PR to add the dependency to `abstract-leveldown` would be welcome too.

There might also be some `process.nextTick`  usage here and there.

Reply:
> So there's no webpack option to inject the buffer polyfill if a module imports it? Because that's how I read "it's possible to manually add a polyfill".

no, the "it's possible to manually add a polyfill" is exactly what we are doing here in this PR.

i will PR abstract-leveldown and process.nextTick if its there it will break my tests and i will PR accordingly.